zain/quiz/README.md

```diff
@@ -0,0 +1,589 @@
+1. What's the output?
+
+function sayHi() {
+  console.log(name);
+  console.log(age);
+  var name = 'Lydia';
+  let age = 21;
+}
+
+sayHi();
+
+A: Lydia and undefined
+
+B: Lydia and ReferenceError
+
+C: ReferenceError and 21
+
+D: ~~undefined and ReferenceError~~
+
+In the sayHi function, both name and age are accessed before they are declared. name is declared with var, so it's hoisted and initialized with undefined, allowing it to be accessed before its actual declaration. However, age is declared with let, and accessing it before its declaration results in a ReferenceError due to the temporal dead zone.
+
+
+2. What's the output?
+
+for (var i = 0; i < 3; i++) {
+  setTimeout(() => console.log(i), 1);
+}
+
+for (let i = 0; i < 3; i++) {
+  setTimeout(() => console.log(i), 1);
+}
+
+A: 0 1 2 and 0 1 2
+
+B: 0 1 2 and 3 3 3
+
+C: ~~3 3 3 and 0 1 2~~
+
+In the first loop, var creates a single variable i with function scope. The asynchronous setTimeout captures the final value of i, which is 3, resulting in 3 3 3 being printed.
+
+In the second loop, let creates a new lexical environment for i in each iteration, resulting in 0 1 2 being printed because each setTimeout callback captures the correct value of i for its respective iteration.
+
+3. What's the output?
+
+const shape = {
+  radius: 10,
+  diameter() {
+    return this.radius * 2;
+  },
+  perimeter: () => 2 * Math.PI * this.radius,
+};
+
+console.log(shape.diameter());
+console.log(shape.perimeter());
+
+A: 20 and 62.83185307179586
+
+B: ~~20 and NaN~~
+
+C: 20 and 63
+
+D: NaN and 63
+
+
+diameter() Method:
+The diameter() method is defined using a regular function within the object shape. This allows this to refer to the object itself, so this.radius correctly accesses the radius property of the shape object.
+perimeter() Method:
+The perimeter property is defined using an arrow function. Arrow functions do not bind their own this, so in this context, this refers to the surrounding scope (which is likely the global scope or undefined). This would lead to an error because this.radius would be undefined.
```

Comment:
when executed in browser this will in this case refer to window object, same is not the case with run times like node.js.